src/main/java/com/cleanroommc/multiblocked/common/capability/trait/ImpetusCapabilityTrait.java

```diff
@@ -0,0 +1,307 @@
+import thecodex6824.thaumicaugmentation.api.impetus.node.NodeHelper;
+import thecodex6824.thaumicaugmentation.api.impetus.node.prefab.BufferedImpetusProsumer;
+import thecodex6824.thaumicaugmentation.api.util.DimensionalBlockPos;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * @author youyihj
+ */
+public class ImpetusCapabilityTrait extends ProgressCapabilityTrait {
+    private ImpetusStorageProxy storage;
+    private ImpetusNodeProxy node;
+    private int capacity;
+    private int maxReceive;
+    private int maxExtract;
+    private int maxInputs;
+    private int maxOutputs;
+    private double beamEndpointXOffset;
+    private double beamEndpointYOffset;
+    private double beamEndpointZOffset;
+
+    public ImpetusCapabilityTrait() {
+        super(ImpetusThaumicAugmentationCapability.CAP);
+    }
+
+    @Override
+    public void serialize(@Nullable JsonElement jsonElement) {
+        super.serialize(jsonElement);
+        if (jsonElement == null) {
+            jsonElement = new JsonObject();
+        }
+        JsonObject jsonObject = jsonElement.getAsJsonObject();
+        capacity = JsonUtils.getInt(jsonObject, "capacity", 300);
+        maxReceive = JsonUtils.getInt(jsonObject, "maxReceive", 50);
+        maxExtract = JsonUtils.getInt(jsonObject, "maxExtract", 50);
+        maxInputs = JsonUtils.getInt(jsonObject, "maxInputs", 1);
+        maxOutputs = JsonUtils.getInt(jsonObject, "maxOutputs", 1);
+        beamEndpointXOffset = JsonUtils.getFloat(jsonObject, "beamEndpointXOffset", 0.5f);
+        beamEndpointYOffset = JsonUtils.getFloat(jsonObject, "beamEndpointYOffset", 0.5f);
+        beamEndpointZOffset = JsonUtils.getFloat(jsonObject, "beamEndpointZOffset", 0.5f);
+        storage = new ImpetusStorageProxy(capacity, maxReceive, maxExtract);
+        node = new ImpetusNodeProxy(maxInputs, maxOutputs, storage);
+    }
+
+    private static Vec3d rotateYCCW(Vec3d toRotate) {
+        double a = toRotate.x - 0.5;
+        double b = toRotate.z - 0.5;
+        return new Vec3d(b + 0.5, toRotate.y, -a + 0.5);
+    }
+
+    @Override
+    public JsonElement deserialize() {
+        JsonObject jsonObject = super.deserialize().getAsJsonObject();
+        jsonObject.addProperty("capacity", capacity);
+        jsonObject.addProperty("maxReceive", maxReceive);
+        jsonObject.addProperty("maxExtract", maxExtract);
+        jsonObject.addProperty("maxInputs", maxInputs);
+        jsonObject.addProperty("maxOutputs", maxOutputs);
+        jsonObject.addProperty("beamEndpointXOffset", beamEndpointXOffset);
+        jsonObject.addProperty("beamEndpointYOffset", beamEndpointYOffset);
+        jsonObject.addProperty("beamEndpointZOffset", beamEndpointZOffset);
+        return jsonObject;
+    }
+
+    @Override
+    protected String dynamicHoverTips(double progress) {
+        return LocalizationUtils.format("multiblocked.gui.trait.impetus.progress", storage.getEnergyStored(), capacity);
+    }
```

Comment:
I'm not sure if there's a problem with the client call? I mean `energyStored` is always synchronized to the client?
if not i suggest that `capacity * progress`